File: leetcode_python/821.py
# 821. 字符的最短距离
# https://leetcode-cn.com/problems/shortest-distance-to-a-character/
from typing import List
from collections import defaultdict
from collections import deque
from math import inf
import bisect
from functools import lru_cache
import heapq
'''
@File : 821.py
@Time : 2022/04/19 22:08:08
@Author : YuMin Zhang
@State : Indepeedent | Half-Depedent | Depedent 
@Thinking :
@Tag : Easy | Medium | Hard
'''
class Solution:
    def shortestToChar(self, s: str, c: str) -> List[int]:
        l = len(s)
        l_dir = [0] * l
        r_dir = [0] * l
        prev = float("-inf")
        for index, val in enumerate(s):
            if val == c:
                prev = index
            l_dir[index] = prev
        prev = float("inf")
        for i in range(l - 1, -1, -1):
            if s[i] == c:
                prev = i
            r_dir[i] = prev
        ans = [0] * l
        for i in range(l):
            ans[i] = min(abs(i - l_dir[i]), abs(i - r_dir[i]))
        return ans

File: leetcode_c_plus_plus/821.cpp
// 821. 字符的最短距离
// https://leetcode-cn.com/problems/shortest-distance-to-a-character/
#include<map>
#include<set>
#include<vector>
#include<cmath>
#include<numeric>
#include<unordered_map>
using namespace std;
#define INT_MAX 2147483647
#define INT_MIN (-INT_MAX - 1)
/**
@File : 821.cpp
@Time : 2022/04/19 22:13:50
@Author : YuMin Zhang
@State : Indepeedent | Half-Depedent | Depedent 
@Thinking :
@Tag : Easy | Medium | Hard
**/
class Solution {
public:
    vector<int> shortestToChar(string s, char c) {
        int n = s.size();
        vector<int> ans(n);
        for(int i = 0, idx = -n; i < n; i++) {
            if(s[i] == c) {
                idx = i;
            }
            ans[i] = i - idx;
        }
        for(int i = n - 1, idx = 2 * n; i >=0; i--) {
            if(s[i] == c) {
                idx = i;
            }
            ans[i] = min(ans[i], idx - i);
        }
        return ans;
    }
};
